File: d19/part1.js
const fs = require('fs');
const readline = require('readline');

const INPUT_FILE = 'in1';

async function processLineByLine() {
  const fileStream = fs.createReadStream(INPUT_FILE);

  const line_reader = readline.createInterface({
    input: fileStream,
    crlfDelay: Infinity
  });

  const blueprints = [];
  for await (const line of line_reader) {
    const split = line.split(".").map(s => s.split(" "));
    blueprints.push([
      new Robot('ore', +split[0][6], 0, 0),
      new Robot('clay', +split[1][5], 0, 0),
      new Robot('obsidian', +split[2][5], +split[2][8], 0),
      new Robot('geode', +split[3][5], 0, +split[3][8]),
    ]);
  }

  const ratings = [];
  for (let i = 0; i < blueprints.length; ++i) {
    ratings.push(rate(blueprints[i], i + 1));
  }

  console.log(ratings.reduce((sum, a) => sum + a));
}

processLineByLine();

function rate(blueprint, blueprintId) {
  let maxGeodes = 0;
  const stack = [new State([1, 0, 0, 0], [0, 0, 0, 0], 24)];

  let minOreCost = Infinity;
  let maxOreCost = 0;
  for (const robot of blueprint) {
    if (robot.oreCost < minOreCost) {
      minOreCost = robot.oreCost;
    }
    if (robot.oreCost > maxOreCost) {
      maxOreCost = robot.oreCost;
    }
  }

  const clayCost = blueprint[2].clayCost;
  const obsidianCost = blueprint[3].obsidianCost;

  while (stack.length > 0) {
    // console.log(stack.length);

    const state = stack.pop();
    
    // console.log(state);
    
    if (state.resources[3] > maxGeodes) {
      maxGeodes = state.resources[3];
    }

    if (state.timeLeft > 0) {
      if (state.resources[0] < maxOreCost + 1 && state.resources[0] >= blueprint[0].oreCost) {
        const newState = state.copy();
        newState.resources[0] -= blueprint[0].oreCost;
        for (let i = 0; i < state.robots.length; ++i) {
          newState.resources[i] += state.robots[i];
        }
        ++newState.robots[0];
        stack.push(newState);
      }

      if (state.resources[1] < clayCost + 1 && state.resources[0] >= blueprint[1].oreCost) {
        const newState = state.copy();
        newState.resources[0] -= blueprint[1].oreCost;
        for (let i = 0; i < state.robots.length; ++i) {
          newState.resources[i] += state.robots[i];
        }
        ++newState.robots[1];
        stack.push(newState);
      }

      if (state.resources[2] < obsidianCost + 1 && state.resources[0] >= blueprint[2].oreCost && state.resources[1] >= blueprint[2].clayCost) {
        const newState = state.copy();
        newState.resources[0] -= blueprint[2].oreCost;
        newState.resources[1] -= blueprint[2].clayCost;
        for (let i = 0; i < state.robots.length; ++i) {
          newState.resources[i] += state.robots[i];
        }
        ++newState.robots[2];
        stack.push(newState);
      }

      if (state.resources[0] >= blueprint[3].oreCost && state.resources[2] >= blueprint[3].obsidianCost) {
        const newState = state.copy();
        newState.resources[0] -= blueprint[3].oreCost;
        newState.resources[2] -= blueprint[3].obsidianCost;
        for (let i = 0; i < state.robots.length; ++i) {
          newState.resources[i] += state.robots[i];
        }
        ++newState.robots[3];
        stack.push(newState);
      }

      if (state.resources[0] < maxOreCost + 1 || state.resources[1] < clayCost + 1 || state.resources[2] < obsidianCost + 1) {
        for (let i = 0; i < state.robots.length; ++i) {
          state.resources[i] += state.robots[i];
        }
        --state.timeLeft;
        stack.push(state);
      }
    }
  }

  console.log(`Blueprint ${blueprintId}: ${maxGeodes}`);
  return maxGeodes * blueprintId;
}

function Robot(type, oreCost, clayCost, obsidianCost) {
  this.type = type;
  this.oreCost = oreCost;
  this.clayCost = clayCost;
  this.obsidianCost = obsidianCost;
}

function State(robots, resources, timeLeft) {
  this.robots = robots;
  this.resources = resources;
  this.timeLeft = timeLeft;

  this.copy = function() {
    return new State([...this.robots], [...this.resources], this.timeLeft - 1);
  }
}

File: d19/part1.py
from copy import copy, deepcopy


def main():
    INPUT_FILE = 'in1'
    
    with open(INPUT_FILE) as f:
        lines = [line.rstrip() for line in f.readlines()]

    blueprints = []
    for line in lines:
        split = [s.split(" ") for s in line.split('.')]
        blueprints.append(
            (
                (int(split[0][6]), 0, 0),
                (int(split[1][5]), 0, 0),
                (int(split[2][5]), int(split[2][8]), 0),
                (int(split[3][5]), 0, int(split[3][8])),
            )
        )
    
    ratings = [rate(blueprint, i + 1) for i, blueprint in enumerate(blueprints)]

    print(sum(ratings))


def rate(blueprint, id):
    max_geodes = 0
    stack = [[[1, 0, 0, 0], [0, 0, 0, 0], 24]]

    max_ore_cost = 0
    for robot in blueprint:
        if robot[0] > max_ore_cost:
            max_ore_cost = robot[0]

    clay_cost = blueprint[2][1]
    obsidian_cost = blueprint[3][2]

    iterations = 0
    while len(stack) > 0:
        iterations += 1
        state = stack.pop()

        if state[1][3] > max_geodes:
            max_geodes = state[1][3]

        potential_max_geodes = state[1][3] + (state[1][3] + state[2]) * state[2] / 2

        if state[2] > 0 and potential_max_geodes > max_geodes:
            if state[0][0] < max_ore_cost and state[1][0] <= max_ore_cost and state[1][0] >= blueprint[0][0]:
                new_state = [copy(state[0]), copy(state[1]), state[2]]
                new_state[1][0] -= blueprint[0][0]
                for i, robots in enumerate(state[0]):
                    new_state[1][i] += robots
                new_state[0][0] += 1
                new_state[2] -= 1
                stack.append(new_state)
            
            if state[0][1] < clay_cost and state[1][1] <= clay_cost and state[1][0] >= blueprint[1][0]:
                new_state = [copy(state[0]), copy(state[1]), state[2]]
                new_state[1][0] -= blueprint[1][0]
                for i, robots in enumerate(state[0]):
                    new_state[1][i] += robots
                new_state[0][1] += 1
                new_state[2] -= 1
                stack.append(new_state)

            if state[0][2] < obsidian_cost and state[1][2] <= obsidian_cost and state[1][0] >= blueprint[2][0] and state[1][1] >= blueprint[2][1]:
                new_state = [copy(state[0]), copy(state[1]), state[2]]
                new_state[1][0] -= blueprint[2][0]
                new_state[1][1] -= blueprint[2][1]
                for i, robots in enumerate(state[0]):
                    new_state[1][i] += robots
                new_state[0][2] += 1
                new_state[2] -= 1
                stack.append(new_state)

            if state[1][0] >= blueprint[3][0] and state[1][2] >= blueprint[3][2]:
                new_state = [copy(state[0]), copy(state[1]), state[2]]
                new_state[1][0] -= blueprint[3][0]
                new_state[1][2] -= blueprint[3][2]
                for i, robots in enumerate(state[0]):
                    new_state[1][i] += robots
                new_state[0][3] += 1
                new_state[2] -= 1
                stack.append(new_state)

            if state[1][0] <= max_ore_cost or state[1][1] <= clay_cost or state[1][2] <= obsidian_cost:
                for i, robots in enumerate(state[0]):
                    state[1][i] += robots
                state[2] -= 1
                stack.append(state)

    print(f'Blueprint {id}: {max_geodes}, iterations: {iterations}')
    return max_geodes * id

if __name__ == '__main__':
    main()